keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,867 @@
+  connectivitiy. There may be multiple implementations, this doc describes
+  expected common behavior.
+
+We propose a new CRD called `ServiceExport`, used to specify which services
+should be exposed across all clusters in the supercluster. `ServiceExports` must
+be created in each cluster that the underlying `Service` resides in. Creation of
+a `ServiceExport` in a cluster will signify that the `Service` with the same
+name and namespace as the export should be visible to other clusters in the
+supercluster.
+
+Another CRD called `ImportedService` will be introduced to store connection
+information about the `Services` in each cluster, e.g. topology. This is
+analogous to the traditional `Service` type in Kubernetes. Each cluster will
+have an `ImportedService` for each `Service` that has been exported within the
+supercluster, referenced by namespaced name.
+
+If multiple clusters export a `Service` with the same namespaced name, they will
+be recognized as a single combined service. The resulting `ImportedService` will
+reference endpoints from both clusters. Properties of the `ImportedService`
+(e.g. ports, topology) will be derived from a merger of component Service
+properties.
+
+Existing implementations of Kubernetes Service API (e.g. kube-proxy) can be
+extended to present `ImportedServices` alongside traditional `Services`.
+
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Different Services Each Deployed to Separate Cluster
+
+I have 2 clusters, each running different services managed by different teams,
+where services from one team depend on services from the other team. I want to
+ensure that a service from one team can discover a service from the other team
+(via DNS resolving to VIP), regardless of the cluster that they reside in. In
+addition, I want to make sure that if the dependent service is migrated to
+another cluster, the dependee is not impacted.
+
+#### Single Service Deployed to Multiple Clusters
+
+I have deployed my stateless service to multiple clusters for redundancy or
+scale. Now I want to propagate topologically-aware service endpoints (local,
+regional, global) to all clusters, so that other services in my clusters can
+access instances of this service in priority order based on availability and
+locality. Requests to my replicated service should seamlessly transition (within
+SLO for dropped requests) between instances of my service in case of failure or
+removal without action by or impact on the caller. Routing to my replicated
+service should optimize for cost metric (e.g.prioritize traffic local to zone,
+region).
+
+```
+<<[UNRESOLVED]>>
+Due to additional constraints that apply to stateful services (e.g. each cluster
```

Comment:
A potential solution here is to allow statefully-named supercluster services. Originally I was thinking cluster name but if you want to support migration it should be something else.

This is similar to a feature to allow arbitrary supercluster names which we could add later (as a field on ExportedService for example). Then the controller would just aggregate based on supercluster name, with the default name being name.namespace.svc.supercluster.local if unspecified.

Reply:
What would a statefully-named supercluster service look like? When you say it would be a field on the export, are you thinking it would be an alias for the service name - as in `exportName: "my-alias"` is exported as `my-alias.namespace.svc.supercluster.local`? That seems like it could solve e.g. export DB-A from cluster A and DB-B from cluster B.

What if I want to run something like Cassandra across multiple clusters? Then I may want a single `my-cassandra.ns.svc.supercluster.local` in front of the aggregate service.  In that instance, it could be solved by requiring that users just use different names (cluster name prefix?) for the stateful set in each cluster, but it would be much better if we could do something that didn't require different configuration in each cluster.

Reply:
Supercluster DNS needs some definition for sure.

To further this discussion, I've been leaning toward this obvious cluster-scoped model. It seems like `<hostname>.<cluster_id>.<svc>.<ns>.*` would provide necessary flexibility. Each pod gets its own unique name supercluster-wide but we still respect the hostname. It does of course mean that you can't connect to a specific pod with only its hostname and service name, but would there be a situation where that's needed and you don't also know the pod's cluster? If you had supercluster-unique hostnames, it would still be possible to look up the pod's FQDN via SRV records

Reply:
Two thoughts:

1) Careful with "cluster ID" vs "cluster name".  There is/was a proposal to formalize ClusterID, and I don't know if that is what you mean.  A "group-unique cluster identifier or name" is probably sufficient.

2) Elsewhere I suggested MAYBE `.<svc>.s.<ns>.<suffix>` to leave room for other kinds of names.  E.g DNS for Gateway is not fully locked down.

Reply:
1. I mean the "group-unique cluster identifier or name", adding cluster_name to the terminology section and use that moving forward so it's clear

2. I think that structure generally makes more sense. My take is that this needs to be driven from Service and MCS can be co-developed to adopt improvements to Service. I'm worried about trying to independently improve DNS for MCS and diverging from Service. <svc>.<ns>.svc.<suffix> is not ideal, but it's consistent and there's no room for confusion coming from single cluster services